pkg/tfshim/sdk-v2/upgrade_state.go

```diff
@@ -6,82 +6,118 @@ import (
 	"strconv"
 
 	"github.com/hashicorp/go-cty/cty"
+	ctyjson "github.com/hashicorp/go-cty/cty/json"
+	"github.com/hashicorp/go-cty/cty/msgpack"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov5"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
 )
 
-func upgradeResourceState(ctx context.Context, p *schema.Provider, res *schema.Resource,
+func upgradeResourceState(ctx context.Context, typeName string, p *schema.Provider, res *schema.Resource,
 	instanceState *terraform.InstanceState) (*terraform.InstanceState, error) {
-
 	if instanceState == nil {
 		return nil, nil
 	}
 
-	m := instanceState.Attributes
-
-	// Ensure that we have an ID in the attributes.
-	m["id"] = instanceState.ID
-
-	version, hasVersion := 0, false
+	version, hasVersion := int64(0), false
 	if versionValue, ok := instanceState.Meta["schema_version"]; ok {
 		versionString, ok := versionValue.(string)
 		if !ok {
 			return nil, fmt.Errorf("unexpected type %T for schema_version", versionValue)
 		}
 		v, err := strconv.ParseInt(versionString, 0, 32)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("failed to parse schema_version: %w", err)
 		}
-		version, hasVersion = int(v), true
+		version, hasVersion = v, true
 	}
```

Comment:
I don't see a check for downgrades here - pretty sure TF doesn't allow downgrades, see https://github.com/pulumi/pulumi-terraform-bridge/pull/1998#discussion_r1607158351